src/proto.rs

```diff
@@ -199,23 +238,115 @@ impl TryFrom<KdcRep> for KerberosAsRep {
         }
 
         let msg_type = KrbMessageType::try_from(rep.msg_type).map_err(|_| {
-            KrbError::InvalidMessageType(rep.msg_type as i32, KrbMessageType::KrbAsRep as i32)
+            KrbError::InvalidEnumValue(
+                std::any::type_name::<KrbMessageType>().to_string(),
+                rep.msg_type as i32,
+            )
+        })?;
+
+        match msg_type {
+            KrbMessageType::KrbAsRep => {
+                let enc_part = EncryptedData::try_from(rep.enc_part)?;
+                trace!(?enc_part);
+
+                let client_realm: String = rep.crealm.into();
+                let client_name: String = rep.cname.into();
+
+                Ok(KerberosAsRep {
+                    client_realm,
+                    client_name,
+                    enc_part,
+                })
+            }
+            _ => Err(KrbError::InvalidMessageType(
+                rep.msg_type as i32,
+                KrbMessageType::KrbAsRep as i32,
+            )),
+        }
+    }
+}
+
+impl TryFrom<KdcRep> for KerberosTgsRep {
+    type Error = KrbError;
+
+    fn try_from(rep: KdcRep) -> Result<Self, Self::Error> {
+        // assert the pvno and msg_type
+        if rep.pvno != 5 {
+            todo!();
+        }
+
+        let msg_type = KrbMessageType::try_from(rep.msg_type).map_err(|_| {
+            KrbError::InvalidEnumValue(
+                std::any::type_name::<KrbMessageType>().to_string(),
+                rep.msg_type as i32,
+            )
         })?;
-        if !matches!(msg_type, KrbMessageType::KrbAsRep) {
+
+        match msg_type {
+            KrbMessageType::KrbTgsRep => Ok(KerberosTgsRep {}),
+            _ => Err(KrbError::InvalidMessageType(
+                rep.msg_type as i32,
+                KrbMessageType::KrbTgsRep as i32,
+            )),
+        }
+    }
+}
+
+impl TryFrom<crate::asn1::krb_error::KrbError> for KerberosErrRep {
+    type Error = KrbError;
+
+    fn try_from(rep: crate::asn1::krb_error::KrbError) -> Result<Self, Self::Error> {
+        // assert the pvno and msg_type
+        if rep.pvno != 5 {
             todo!();
         }
 
-        let enc_part = EncryptedData::try_from(rep.enc_part)?;
-        trace!(?enc_part);
+        let msg_type = KrbMessageType::try_from(rep.msg_type).map_err(|_| {
+            KrbError::InvalidEnumValue(
+                std::any::type_name::<KrbMessageType>().to_string(),
+                rep.msg_type as i32,
+            )
+        })?;
+
+        match msg_type {
+            KrbMessageType::KrbError => {
+                let error_code = KrbErrorCode::try_from(rep.error_code).map_err(|_| {
+                    KrbError::InvalidEnumValue(
+                        std::any::type_name::<KrbErrorCode>().to_string(),
```

Comment:
We already bubble up the error code itself, so we can probably omit the string here.